rotkehlchen/tests/utils/blockchain.py

```diff
@@ -670,27 +666,32 @@ def mock_base_has_activity(account):
     stack.enter_context(patch.object(
         chains_aggregator.optimism.node_inquirer.etherscan,
         'has_activity',
-        side_effect=mock_optimism_has_activity,
+        side_effect=lambda account: mock_chain_has_activity(account, SupportedBlockchain.OPTIMISM),
```

Comment:
In the same spirit as above. Do a for loop here and until the bottom. This is how abstraction shoudl work. Don't just add copy paste lines. (Similar can happen in many other places -- but we can find them as we go)

```python
for chain in EVM_CHAINS_WITH_TRANSACTIONS:
    stack.enter_context(patch.object(
        chains_aggregator.get_evm_manager(chain.chain_id()).node_inquirer,
        'has_activity',
        side_effect=lambda account: mock_chain_has_activity(account, chain),
    ))

Reply:
Both of the following approaches don't seem to work. Probably the target of patch.object needs to be exactly as it is called in the mocked function. Needs a bit more thinking on how this should be done. For not delaying the rest of the review I leave it as it is for now. 

```
for chain in EVM_CHAINS_WITH_TRANSACTIONS:
    stack.enter_context(patch.object(
        getattr(chains_aggregator, chain.to_chain_id().to_name()).node_inquirer.etherscan,
        'has_activity',
        side_effect=lambda account: mock_chain_has_activity(account, chain),
    ))
```

```
for chain in EVM_CHAINS_WITH_TRANSACTIONS:
    stack.enter_context(patch.object(
        chains_aggregator.get_evm_manager(chain.to_chain_id()).node_inquirer.etherscan,
        'has_activity',
        side_effect=lambda account: mock_chain_has_activity(account, chain),
    ))
```

Reply:
This is a basic python mistake. The reason it did not work for you is the lambda. A lambda forms a clojure and unless specific bindings are specified takes the last value of the arguments passed. Look here to see how common a mistake this is: https://stackoverflow.com/questions/10724495/getting-all-arguments-and-values-passed-to-a-function

Good thing is it's easily fixable.

Do this `side_effect=lambda account, i_chain=chain: mock_chain_has_activity(account, i_chain),` to ind the chain for each iteration.

I tested the whole thing, combined with the locals() suggestion above. The diff to make it work is below

```diff
modified   rotkehlchen/tests/utils/blockchain.py
@@ -22,6 +22,7 @@ from rotkehlchen.serialization.deserialize import deserialize_evm_address
 from rotkehlchen.tests.utils.eth_tokens import CONTRACT_ADDRESS_TO_TOKEN
 from rotkehlchen.tests.utils.mock import MockResponse
 from rotkehlchen.types import (
+    EVM_CHAINS_WITH_TRANSACTIONS,
     SUPPORTED_EVM_CHAINS,
     BTCAddress,
     ChecksumEvmAddress,
@@ -610,23 +611,15 @@ def setup_evm_addresses_activity_mock(
         stack: 'ExitStack',
         chains_aggregator: 'ChainsAggregator',
         eth_contract_addresses: list[ChecksumEvmAddress],
-        ethereum_addresses: list[ChecksumEvmAddress],
+        ethereum_addresses: list[ChecksumEvmAddress],  # pylint: disable=unused-argument  # used by the saved locals  # noqa: E501, RUF100
         avalanche_addresses: Optional[list[ChecksumEvmAddress]] = None,
-        optimism_addresses: Optional[list[ChecksumEvmAddress]] = None,
-        polygon_pos_addresses: Optional[list[ChecksumEvmAddress]] = None,
-        arbitrum_one_addresses: Optional[list[ChecksumEvmAddress]] = None,
-        base_addresses: Optional[list[ChecksumEvmAddress]] = None,
-        gnosis_addresses: Optional[list[ChecksumEvmAddress]] = None,
+        optimism_addresses: Optional[list[ChecksumEvmAddress]] = None,  # pylint: disable=unused-argument  # used by the saved locals  # noqa: E501, RUF100
+        polygon_pos_addresses: Optional[list[ChecksumEvmAddress]] = None,  # pylint: disable=unused-argument  # used by the saved locals  # noqa: E501, RUF100
+        arbitrum_one_addresses: Optional[list[ChecksumEvmAddress]] = None,  # pylint: disable=unused-argument  # used by the saved locals  # noqa: E501, RUF100
+        base_addresses: Optional[list[ChecksumEvmAddress]] = None,  # pylint: disable=unused-argument  # used by the saved locals  # noqa: E501, RUF100
+        gnosis_addresses: Optional[list[ChecksumEvmAddress]] = None,  # pylint: disable=unused-argument  # used by the saved locals  # noqa: E501, RUF100
 ) -> 'ExitStack':
-    mocked_addresses = {
-        'ethereum_addresses': ethereum_addresses,
-        'avalanche_addresses': avalanche_addresses,
-        'optimism_addresses': optimism_addresses,
-        'polygon_pos_addresses': polygon_pos_addresses,
-        'arbitrum_one_addresses': arbitrum_one_addresses,
-        'base_addresses': base_addresses,
-        'gnosis_addresses': gnosis_addresses,
-    }
+    saved_locals = locals()  # bit hacky, but save locals here so they can be accessed by mock_chain_has_activity  # noqa: E501
 
     def mock_ethereum_get_code(account):
         if account in eth_contract_addresses:
@@ -644,7 +637,7 @@ def setup_evm_addresses_activity_mock(
         return ZERO
 
     def mock_chain_has_activity(account: ChecksumEvmAddress, chain: SUPPORTED_EVM_CHAINS):
-        addresses = mocked_addresses[f'{chain.to_chain_id().to_name()}_addresses']
+        addresses = saved_locals[f'{chain.to_chain_id().to_name()}_addresses']
         return EtherscanHasChainActivity.TRANSACTIONS if addresses is not None and account in addresses else EtherscanHasChainActivity.NONE  # noqa: E501
 
     stack.enter_context(patch.object(
@@ -662,36 +655,14 @@ def setup_evm_addresses_activity_mock(
         'get_avax_balance',
         side_effect=mock_avax_balance,
     ))
-    stack.enter_context(patch.object(
-        chains_aggregator.optimism.node_inquirer.etherscan,
-        'has_activity',
-        side_effect=lambda account: mock_chain_has_activity(account, SupportedBlockchain.OPTIMISM),
-    ))
-    stack.enter_context(patch.object(
-        chains_aggregator.polygon_pos.node_inquirer.etherscan,
-        'has_activity',
-        side_effect=lambda account: mock_chain_has_activity(account, SupportedBlockchain.POLYGON_POS),  # noqa: E501
-    ))
-    stack.enter_context(patch.object(
-        chains_aggregator.ethereum.node_inquirer.etherscan,
-        'has_activity',
-        side_effect=lambda account: mock_chain_has_activity(account, SupportedBlockchain.ETHEREUM),
-    ))
-    stack.enter_context(patch.object(
-        chains_aggregator.arbitrum_one.node_inquirer.etherscan,
-        'has_activity',
-        side_effect=lambda account: mock_chain_has_activity(account, SupportedBlockchain.ARBITRUM_ONE),  # noqa: E501
-    ))
-    stack.enter_context(patch.object(
-        chains_aggregator.base.node_inquirer.etherscan,
-        'has_activity',
-        side_effect=lambda account: mock_chain_has_activity(account, SupportedBlockchain.BASE),
-    ))
-    stack.enter_context(patch.object(
-        chains_aggregator.gnosis.node_inquirer.etherscan,
-        'has_activity',
-        side_effect=lambda account: mock_chain_has_activity(account, SupportedBlockchain.GNOSIS),
-    ))
+
+    for chain in EVM_CHAINS_WITH_TRANSACTIONS:
+        stack.enter_context(patch.object(
+            chains_aggregator.get_evm_manager(chain.to_chain_id()).node_inquirer.etherscan,
+            'has_activity',
+            side_effect=lambda account, i_chain=chain: mock_chain_has_activity(account, i_chain),
+        ))
+
     return stack
```